### PR TITLE
SI-8341 Refine handoff of undet. params from implicit search

### PR DESCRIPTION
**Pull request by retronym**
In SI-7944 / 251c2b93, we discovered that typechecking of an implicit
candidate could leave orphaned undetermined type parameters in the
implicit search context. This resulted in naked type parameters
leaking into implicit expansions. The fix seemed easy: just copy
any symbols from `implicitSearchContext.undetparams` to the enclosing
context (other than ones in `SearchResult#subst`).

However, the test case in this ticket reveals a subtle flaw in that
fix: `implicitSerachContext.undetparams` only contains the type params
from the most recently typechecked candidate!

Why? Implicit search uses the same context to typecheck all plausibly
compatible candidates.  The typechecking itself is driven by
`typedImplicit1`.

Side note, that explains the heisenbug behaviour noted in the ticket:
Not _all_ plausibly implicit candidates are typechecked. If
the current 'best' eligible candidate is more specific than the next
candidate, we can skip that altogether.Implicit search actually
exploits this for performance by ordering the candidates according to
usage statistics. This reordering, means that commenting out lines
elsewhere in the file changed the behaviour!

This commit simply stores the undet. tparams in the `SearchResult`,
where it is safe from the vaguries of typechecking other candidates.

That makes `Test1` and `Test2` with in the enclosed test case fail
uniformly, both with each other, and with an explicit call to
the view.

This is ostensibly a regression from 2.10.3. To get there,
we need an implicit search that has to infer `Nothing` in a covariant
position. In 2.10.3, we would just let the `G` out into the wild,
which did the right thing for the wrong reasons.

### CODE REVIEW

**Comment by adriaanm** [comment]
Beautiful analysis, @retronym.

File: test/files/neg/t8431.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+class TestExplicit {
+  import C.{cbf, convert2}
+
+  val s: Invariant[Nothing] = ???
+
+  // Now the implicit Test fail uniformly as per this explicit conversion
+  convert2(s).combined
+
+  // Breaking this expression down makes it work.
```
Comment (retronym) [inline test/files/neg/t8431.scala:47]: This comment is wrong

Comment (retronym) [inline test/files/neg/t8431.scala:47]: But I'll let the build run before changing this.

**Comment by retronym** [comment]
I've scheduled a dbuild of this: https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/ (build 8, pending)

**Comment by adriaanm** [comment]
lets fix the comment in a follow-up commit. Otherwise, LGTM

**Comment by adriaanm** [comment]
Success! https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/8/console

**Comment by adriaanm** [comment]
It would be so cool to have a bot to launch and watch jenkins jobs...

File: test/files/neg/t8431.scala (hunk 1 above)
Comment (adriaanm) [inline test/files/neg/t8431.scala:47]: @retronym, based on the checkfile, I assume the comment is missing a negation: "... does not make it work."
I'll merge this PR and patch the check file in a follow-up PR unless you beat me to it in the next 20 min or so.

**Comment by adriaanm** [comment]
LGTM, merging and submitting fixup PR in a minute

**Pull request merged**